version 1.0.0, closes #6

IMax153 opened the pull request:
Hello @gcanti! As discussed in #6, I have ported `graphics-ts` over to `fp-ts` `v2.x`. There were quite a few major changes that I made during the port. I have tried to list the majority of them below.

- Upgraded all dependencies to their latest versions
- Upgraded/added several configuration files in the repository
  - I tried to follow the configuration that you have set up in the `fp-ts` repo as much as possible.
- Added a test suite 
  - 100% coverage of all modules
- Modernized the `example` directory to run off of `webpack-dev-server` with HMR 
  - Allows those cloning the repo to observe their changes to the examples in real time
- Removed the `free-canvas` example from the `example` directory 
  - Replaced with an example of a clipped rect from the [MDN documentation](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/clip)
- See the [CHANGELOG.md](https://github.com/IMax153/graphics-ts/blob/1.x/CHANGELOG.md)  for all module-related changes

I did not bump the package version or adjust the CI configuration. I figured that you could bump the version when publishing, and update the CI config if you feel it is necessary.

Let me know if there are any additional changes you would like me to make! One thing I had though about was to add `docs-ts` support to the repository, but I figured I would see if you were interested in that first!

IMax153 committed:
Update configuration files and package dependencies

IMax153 committed:
Update the Color module

IMax153 committed:
Update the Font module

IMax153 committed:
Separate shapes into their own Shape module

IMax153 committed:
Update the Canvas module and prettify all files

IMax153 committed:
Update the Drawing module

IMax153 committed:
Setup testing environment

IMax153 committed:
Add tests for the Color module

IMax153 committed:
Add tests for the Font module

IMax153 committed:
Add tests for the Shapes module

IMax153 committed:
Add tests for the Canvas module

IMax153 committed:
Add tests for the Drawing module and rename Shapes module to Shape

gcanti commented on src/Canvas.ts:
```diff
@@ -0,0 +1,1011 @@
+ *
+ * @since 1.0.0
+ */
+export const setWidth: (width: number) => (canvas: HTMLCanvasElement) => IO.IO<void> = (w) => (c) => () => {
```
We could return `HTMLCanvasElement` instead of `void`

```diff
- export const setWidth: (width: number) => (canvas: HTMLCanvasElement) => IO.IO<void> = (w) => (c) => () => {
+export const setWidth: (width: number) => (canvas: HTMLCanvasElement) => IO.IO<HTMLCanvasElement> = (w) => (c) => () => {
  c.width = w
+  return c
}
```

this will simplify `setDimensions`

```ts
export const setDimensions: (dimensions: CanvasDimensions) => (canvas: HTMLCanvasElement) => IO.IO<HTMLCanvasElement> = (d) => (c) =>
  pipe(
    c,
    setWidth(d.width),
    IO.chain(setHeight(d.height))
  )
```

IMax153 replied:
Makes sense to me. See https://github.com/gcanti/graphics-ts/pull/7/commits/7d16456374e7b1649daca1088a9035c88a3605dd

gcanti commented on src/Canvas.ts:
```diff
@@ -0,0 +1,1011 @@
+ *
+ * @since 1.0.0
+ */
+export const setHeight: (height: number) => (canvas: HTMLCanvasElement) => IO.IO<void> = (h) => (c) => () => {
```
same comment as `setWidth`

IMax153 replied:
https://github.com/gcanti/graphics-ts/pull/7/commits/7d16456374e7b1649daca1088a9035c88a3605dd

gcanti commented on src/Canvas.ts:
```diff
@@ -0,0 +1,1011 @@
+ *
+ * @since 1.0.0
+ */
+export const setDimensions: (dimensions: CanvasDimensions) => (canvas: HTMLCanvasElement) => IO.IO<void> = (d) => (c) =>
```
idem, should return `IO.IO<HTMLCanvasElement>`

IMax153 replied:
https://github.com/gcanti/graphics-ts/pull/7/commits/7d16456374e7b1649daca1088a9035c88a3605dd

gcanti commented on src/Drawing.ts:
````diff
@@ -0,0 +1,654 @@
+ * Gets a `Monoid` instance for `OutlineStyle`.
+ *
+ * @example
+ * ```typescript
````
you can remove the fence

```diff
 * @example
- * ```typescript
 * import * as O from 'fp-ts/lib/Option';
 * import * as C from 'graphics-ts/lib/Color';
 * import * as D from 'graphics-ts/lib/Drawing';
```

IMax153 replied:
Didn't realize these were not necessary! Thanks 👍. See https://github.com/gcanti/graphics-ts/pull/7/commits/aaf3a44f1d61e72ab05ea1e2a41064c86089f2e1

gcanti commented on src/Drawing.ts:
```diff
@@ -0,0 +1,654 @@
+  const go: (drawing: Drawing) => (ctx: CanvasRenderingContext2D) => IO.IO<CanvasRenderingContext2D> = (drawing) => {
+    switch (drawing._tag) {
+      case 'Clipped':
+        return flow(
```
There are some weird usages of `flow` (with only one argument) in this function

I mean

```
flow(f) = f
```

IMax153 replied:
Apologies, these must have been left over from when I was trying to clean up the nested arrow functions. I think I removed all of the extraneous calls to `flow`. See https://github.com/gcanti/graphics-ts/pull/7/commits/aaf3a44f1d61e72ab05ea1e2a41064c86089f2e1

gcanti commented on src/Font.ts:
```diff
@@ -0,0 +1,159 @@
+ *
+ * @since 1.0.0
+ */
+export const ShowFontOptions: S.Show<FontOptions> = {
```
following the usual naming convention, should be named `showFontOptions`

IMax153 replied:
No problem, I should have probably followed the naming convention from the start. See https://github.com/gcanti/graphics-ts/pull/7/commits/7898462206775283d6538be112aef83ceac2e28f

gcanti commented on src/Font.ts:
```diff
@@ -0,0 +1,159 @@
+ *
+ * @since 1.0.0
+ */
+export const showFontOptions = ShowFontOptions.show
```
I wouldn't export a `showFontOptions` function, there's `S.Show<FontOptions>` for that

IMax153 replied:
See https://github.com/gcanti/graphics-ts/pull/7/commits/7898462206775283d6538be112aef83ceac2e28f

gcanti commented on src/Font.ts:
```diff
@@ -0,0 +1,159 @@
+ *
+ * @since 1.0.0
+ */
+const ShowFont: S.Show<Font> = {
```
idem, should be `showFont`

IMax153 replied:
See https://github.com/gcanti/graphics-ts/pull/7/commits/7898462206775283d6538be112aef83ceac2e28f

gcanti commented on src/Font.ts:
```diff
@@ -0,0 +1,159 @@
+ *
+ * @since 1.0.0
+ */
+export const showFont = ShowFont.show
```
idem, I wouldn't export a `showFont` function

IMax153 replied:
See https://github.com/gcanti/graphics-ts/pull/7/commits/7898462206775283d6538be112aef83ceac2e28f

gcanti commented:
@IMax153 wow, great job!

> I did not bump the package version or adjust the CI configuration. I figured that you could bump the version when publishing, and update the CI config if you feel it is necessary.

ok

> One thing I had though about was to add docs-ts support to the repository, but I figured I would see if you were interested in that first!

Adding `docs-ts@latest` would be nice (it outputs a web site containing the API reference)

IMax153 committed:
Return canvas element from dimension getter/setter

Updates `getWidth`, `setWidth`, and `setDimensions` to return
IO.IO<HTMLCanvasElement> instead of void.

(see https://github.com/gcanti/graphics-ts/pull/7#discussion_r428113520)

gcanti commented on src/Canvas.ts:
```diff
@@ -0,0 +1,1015 @@
+  pipe(
+    save(ctx),
+    IO.chain(() => f(ctx)),
+    IO.chain((a) =>
+      pipe(
+        restore(ctx),
+        IO.map(() => a)
+      )
+    )
```
this pattern which is a bit cumbersome...

```ts
IO.chain((a) =>
  pipe(
    restore(ctx),
    IO.map(() => a)
  )
)
```

...can be replaced by `chainFirst`

```ts
IO.chainFirst(() => restore(ctx))
```

IMax153 replied:
I discover cleaner/better ways to write code with `fp-ts` every day! Thanks for the tip @gcanti. See https://github.com/gcanti/graphics-ts/pull/7/commits/375baca3c6c6687cd48c75c850d3efcbfc8ac92b

IMax153 committed:
Remove weird usage of flow from render.

See (https://github.com/gcanti/graphics-ts/pull/7#discussion_r428116696)

Also removes unnecessary fence from monoidOutlineStyle example.

See (https://github.com/gcanti/graphics-ts/pull/7#discussion_r428115239)

IMax153 committed:
Follow naming conventions for Show instances.

Removes the shortcut functions to the show instances for Font
and FontOptions.

See (https://github.com/gcanti/graphics-ts/pull/7#discussion_r428117400)

See (https://github.com/gcanti/graphics-ts/pull/7#discussion_r428117760)

See (https://github.com/gcanti/graphics-ts/pull/7#discussion_r428118179)

IMax153 committed:
Remove unnecessary index.ts file from tests.

IMax153 committed:
Convert to IO.chainFirst where possible

See (https://github.com/gcanti/graphics-ts/pull/7#discussion_r428166487)

IMax153 commented:
Alright, I believe that I have made all the changes requested. Let me know if there is anything else that you come across!

> Adding `docs-ts@latest` would be nice (it outputs a web site containing the API reference)

I will see if I can get `docs-ts@latest` working - I will let you know once I have tested it out.

IMax153 committed:
Make fp-ts a peer dependency of graphics-ts

IMax153 committed:
Add initial documentation to the library

IMax153 committed:
Adjust install documentation in README.

IMax153 committed:
Remove old duplicate install documentation

IMax153 commented:
@gcanti - Added in the documentation using `docs-ts`

Are there any additional changes that you think should be made?

gcanti commented:
> Are there any additional changes that you think should be made?

There are some usages of `IO.of` that don't seem right, example:

```ts
// Canvas.ts
export const getWidth: (canvas: HTMLCanvasElement) => IO.IO<number> = (c) => IO.of(c.width)
```

should be

```ts
export const getWidth: (canvas: HTMLCanvasElement) => IO.IO<number> = (c) => () => c.width
```

`IO.of` lifts a pure value into `IO`, but here we are reading from the `HTMLCanvasElement`, which is a side effect.
Otherwise you end up reading from `HTMLCanvasElement` when you call `getWidth` instead of when the `IO` action will be executed.

IMax153 committed:
Replace impure calls to IO.of with thunks

IMax153 commented:
Ah, thank you @gcanti for the clarification, this makes a lot more sense to me now. 

I was not thinking about the fact that the value `a` passed to `IO.of(a)` is actually evaluated immediately when `IO.of` is called, even though it is then passed to a thunk within `IO.of`. 

I made the fix in https://github.com/gcanti/graphics-ts/pull/7/commits/c002bc3d454a1e62c4101f69a9a96899b0c67bed.

gcanti commented on src/Drawing.ts:
```diff
@@ -572,7 +572,7 @@ const applyStyle: <A>(
     fa,
     O.fold(
-      () => IO.of(ctx),
+      () => () => ctx,
```
actually `IO.of(ctx)` was ok here

IMax153 replied:
See https://github.com/gcanti/graphics-ts/pull/7/commits/d3848bcfd2b47782484a1da84fe31d5e2cae11f8

gcanti commented on src/Drawing.ts:
```diff
@@ -586,15 +586,15 @@ const renderShape: (shape: Shape) => (ctx: CanvasRenderingContext2D) => IO.IO<Ca
         pipe(
           readOnlyArrayTraverseIO(shape.shapes, (s) => pipe(ctx, renderShape(s))),
-          IO.chain(() => IO.of(ctx))
+          IO.chain(() => () => ctx)
```
this is

```ts
IO.map(() => ctx)
```

IMax153 replied:
See https://github.com/gcanti/graphics-ts/pull/7/commits/d3848bcfd2b47782484a1da84fe31d5e2cae11f8

gcanti commented on src/Drawing.ts:
```diff
@@ -586,15 +586,15 @@ const renderShape: (shape: Shape) => (ctx: CanvasRenderingContext2D) => IO.IO<Ca
           shape.points,
           ROA.foldLeft(
-            () => IO.of(ctx),
+            () => () => ctx,
```
`IO.of(ctx)` was ok

IMax153 replied:
See https://github.com/gcanti/graphics-ts/pull/7/commits/d3848bcfd2b47782484a1da84fe31d5e2cae11f8

gcanti commented on src/Drawing.ts:
```diff
@@ -604,7 +604,7 @@ const renderShape: (shape: Shape) => (ctx: CanvasRenderingContext2D) => IO.IO<Ca
                     shape.closed,
                     B.fold(
-                      () => IO.of(ctx),
+                      () => () => ctx,
```
`IO.of(ctx)` was ok

IMax153 replied:
See https://github.com/gcanti/graphics-ts/pull/7/commits/d3848bcfd2b47782484a1da84fe31d5e2cae11f8

gcanti commented on src/Drawing.ts:
```diff
@@ -645,7 +645,7 @@ export const render: (drawing: Drawing) => (ctx: CanvasRenderingContext2D) => IO
           pipe(
             readOnlyArrayTraverseIO(drawing.drawings, (d) => go(d)(ctx)),
-            IO.chain(() => IO.of(ctx))
+            IO.chain(() => () => ctx)
```
this is

```ts
IO.map(() => ctx)
```

IMax153 replied:
See https://github.com/gcanti/graphics-ts/pull/7/commits/d3848bcfd2b47782484a1da84fe31d5e2cae11f8

IMax153 committed:
Revert erroneous removals of IO.of
Also use IO.map instead of IO.chain where applicable

See (https://github.com/gcanti/graphics-ts/pull/7#discussion_r429560380
, https://github.com/gcanti/graphics-ts/pull/7#discussion_r429560433,
https://github.com/gcanti/graphics-ts/pull/7#discussion_r429560454,
https://github.com/gcanti/graphics-ts/pull/7#discussion_r429560462,
https://github.com/gcanti/graphics-ts/pull/7#discussion_r429560549)

IMax153 commented:
@gcanti Thank you for your patience. I am still learning how to navigate `fp-ts` as well as FP concepts in general, so I appreciate your review and comments.

gcanti commented:
@IMax153 No problem. 

The code is ok but...

> Your libraries/blog posts have inspired me to learn more about Category Theory as well as the functional programming paradigm

...as a learning exercise I would propose another change. 

In `Canvas.ts` you can see that almost all functions have the following signature

**HTMLCanvasElement**

```ts
(canvas: HTMLCanvasElement) => IO.IO<A>
```

**CanvasRenderingContext2D**

```ts
(ctx: CanvasRenderingContext2D) => IO.IO<CanvasRenderingContext2D>
```

Basically we are _reading_ from `HTMLCanvasElement` (or `CanvasRenderingContext2D`) and returning a type `A` wrapped in `IO`.

So we could say that when managing `HTMLCanvasElement` we are yielding a "HTML effect"

```ts
import * as R from 'fp-ts/lib/Reader'

export interface HTML<A> extends R.Reader<HTMLCanvasElement, IO.IO<A>> {}
```

when managing `CanvasRenderingContext2D` we are yielding a "Render effect"

```ts
export interface Render<A> extends R.Reader<CanvasRenderingContext2D, IO.IO<A>> {}
```

and when managing `CanvasGradient` we are yielding a "Gradient effect"

```ts
export interface Gradient<A> extends R.Reader<CanvasGradient, IO.IO<A>> {}
```

so this

```ts
export const getContext2D: (canvas: HTMLCanvasElement) => IO.IO<CanvasRenderingContext2D> = (c) =>
  IO.of(unsafeGetContext2D(c))

export const getWidth: (canvas: HTMLCanvasElement) => IO.IO<number> = (c) => () => c.width
// etc...
```

becomes this

```ts
export const getContext2D: HTML<CanvasRenderingContext2D> = (c) => IO.of(unsafeGetContext2D(c))

export const getHeight: HTML<number> = (c) => () => c.width
// etc...
```

and this

```ts
export const setLineWidth: (width: number) => (ctx: CanvasRenderingContext2D) => IO.IO<CanvasRenderingContext2D> = (
  w
) => (ctx) => () => {
  ctx.lineWidth = w
  return ctx
}

export const strokePath: <A>(
  f: (ctx: CanvasRenderingContext2D) => IO.IO<A>
) => (ctx: CanvasRenderingContext2D) => IO.IO<A> = (f) => (ctx) =>
  pipe(
    ctx,
    beginPath,
    IO.chain(() => f(ctx)),
    IO.chain((a) =>
      pipe(
        ctx,
        stroke(),
        IO.map(() => a)
      )
    )
  )

// etc...  
```

becomes this

```ts
export const setLineWidth: (width: number) => Render<CanvasRenderingContext2D> = (w) => (ctx) => () => {
  ctx.lineWidth = w
  return ctx
}

export const strokePath: <A>(f: Render<A>) => Render<A> = (f) => (ctx) =>
  pipe(
    ctx,
    beginPath,
    IO.chain(() => f(ctx)),
    IO.chain((a) =>
      pipe(
        ctx,
        stroke(),
        IO.map(() => a)
      )
    )
  )
```

This change will make the signature way more readable.

IMax153 committed:
Model canvas functions as Readers which yield an effect

IMax153 committed:
Update generated docs

IMax153 commented:
@gcanti This change certainly makes the function signatures more readable. I also feel like representing the `Canvas` module functions as `Reader`s that return an effectful computation makes more sense semantically, especially for newcomers to the library who are looking to understand how the library functions.

I am trying to develop intuition for when and where I can make improvements like this in my own code, but it is difficult given my lack of experience with FP. In addition to continuing to practice utilizing these concepts, do you by any chance have any resources, whether they be textbooks, blogs, video series, etc., that you would recommend for someone like myself who is looking to continue develop their FP skills?

gcanti commented:
@IMax153 check this out https://twitter.com/marcusradell/status/1262272551222489095

gcanti commented:
> This change certainly makes the function signatures more readable

Nice, but there's more. Look at this function

```ts
export const strokePath: <A>(f: Render<A>) => Render<A> = (f) => (ctx) =>
  pipe(
    ctx,
    beginPath,
    IO.chain(() => f(ctx)),
    IO.chainFirst(() => pipe(ctx, stroke()))
  )
```

You are threading the context (`ctx`) by hand.

What if you could just chain the three `Render` effects?

- `beginPath`
- `f`
- `stroke()`

We would need a `Monad` instance of `Render`. Does `Render` admit a `Monad` instance?
Sure it does! We know this because theory teels us that

```ts
Reader<R, M<A>>
```

admits a `Monad` instance for any effect `M` as long as `M` admits a `Monad` instance.
In out case we have

```ts
Reader<R, IO<A>>
```

and, since `IO` admits a `Monad` instance, we know that `Render<A> = Reader<CanvasRenderingContext2D, IO<A>>` admits a `Monad` instance too.

We can import `ReaderIO<R, A> = Reader<R, IO<A>>` from `fp-ts-contrib` and change the `Render` definition like this

```ts
import * as R from 'fp-ts-contrib/lib/ReaderIO'

export interface Render<A> extends R.ReaderIO<CanvasRenderingContext2D, A> {}
```

Now we can refactor `strokePath`

```ts
export const strokePath: <A>(f: Render<A>) => Render<A> = (f) =>
  pipe(
    beginPath,
    R.chain(() => f),
    R.chainFirst(() => stroke())
  )
```

Another refactoring example, from

```ts
export const withContext: <A>(f: Render<A>) => Render<A> = (f) => (ctx) =>
  pipe(
    save(ctx),
    IO.chain(() => f(ctx)),
    IO.chainFirst(() => restore(ctx))
  )
```

to

```ts
export const withContext: <A>(f: Render<A>) => Render<A> = (f) =>
  pipe(
    save,
    R.chain(() => f),
    R.chainFirst(() => restore)
  )
```

Another refactoring (this is the example contained in the `Canvas.ts` module documentation)

```ts
import * as R from 'fp-ts-contrib/lib/ReaderIO'
import { error } from 'fp-ts/lib/Console'
import { flow } from 'fp-ts/lib/function'
import * as IO from 'fp-ts/lib/IO'
import * as O from 'fp-ts/lib/Option'
import { pipe } from 'fp-ts/lib/pipeable'
import * as C from '../src/Canvas'
import * as Color from '../src/Color'
import * as S from '../src/Shape'

const render = (canvasId: string) => <A>(r: C.Render<A>): IO.IO<void> =>
  pipe(
    C.getCanvasElementById(canvasId),
    IO.chain(O.fold(() => error(`[ERROR]: Unable to find canvas`), flow(C.getContext2D, IO.chain(r))))
  )

const triangle = C.fillPath(
  pipe(
    C.setFillStyle(pipe(Color.black, Color.toCss)),
    R.chain(() => C.moveTo(S.point(75, 50))),
    R.chain(() => C.lineTo(S.point(100, 75))),
    R.chain(() => C.lineTo(S.point(100, 25)))
  )
)

render('canvas')(triangle)()
```

IMax153 committed:
Refactor Canvas module to use ReaderIO

See discussion -
(https://github.com/gcanti/graphics-ts/pull/7#issuecomment-633808605)

IMax153 committed:
Refactor Drawing module to utilize ReaderIO

IMax153 committed:
Refactor imports to follow convention

IMax153 committed:
Update example and docs to use ReaderIO

IMax153 commented:
@gcanti - Thank you very much once again for demonstrating how the code can be made more readable, as well as the explanation of the theory behind the changes - it certainly helps me better understand how composition can be used to abstract away details and improve the declarative nature/readability of the code.

I went about searching for places that the code could be refactored using the `ReaderIO` monad. One place that I did get stuck on is the pattern below that emerged in the `Drawing` module. 

```typescript
// Drawing.tx (renderShape)
case 'Composite':
  return (ctx) =>
    pipe(
      ctx,
      traverseReaderIO(shape.shapes, renderShape),
      IO.map(() => ctx)
    )
```

and here

```typescript
// Drawing.ts (render)
case 'Many':
  return (ctx) =>
    pipe(
      ctx,
      traverseReaderIO(d.drawings, go),
      IO.map(() => ctx)
    )
```

Each time that I use `traverseReaderIO`, I am forced to pipe the `CanvasRenderingContext2D` through to ensure that I can use `IO.map` to return a single `CanvasRenderingContext2D` as opposed to the `Array<CanvasRenderingContext2D>` returned by `traverseReaderIO`. I am sure that this is an anti-pattern, and was hoping that you could provide some guidance.

gcanti commented:
You could use `ask`

```ts
case 'Composite':
  return pipe(
    traverseReaderIO(shape.shapes, renderShape),
    R.chain(() => R.ask())
  )
```

IMax153 committed:
Use R.ask() to avoid piping ctx

IMax153 commented:
Perfect! Thank you!

IMax153 committed:
Refactor Canvas module tests to use ReaderIO

IMax153 committed:
Refactor Drawing module tests to use ReaderIO

IMax153 commented:
Hey @gcanti - I worked on refactoring the tests a bit to make use of the updates to the `Canvas` and `Drawing` modules. I figured that users of the library may refer to the tests to gain a better understanding of how `graphics-ts` functions, so they should probably represent intended use. 

I made heavy use of the the helper function below.

```typescript
const render: <A>(fa: C.Render<A>) => IO.IO<A> = (fa) => pipe(canvas, C.getContext2D, IO.chain(fa))
```

One problem that I am having is how I can thread `Gradient<A>` effects into `Render<A>` effects. This is what I have now, but I feel like both myself and those referring to the tests for examples would benefit from a better example.

```typescript
pipe(
  canvas,
  C.getContext2D,
  IO.chain((ctx) =>
    pipe(
      gradient,
      C.addColorStop(0, acqua),
      IO.chain(C.addColorStop(0.5, green)),
      IO.chain(C.addColorStop(1, green)),
      IO.chain((g) => pipe(ctx, C.setFillStyle(g), IO.chain(C.fillRect(rect))))
    )
  )
)()
```

Another question I had is whether we should export a helper function for querying the DOM for a canvas and executing a `Render<A>` effect, similar to the `render` function that we created in the `example` directory. 

Perhaps this would be abstracting to much away and getting a bit verbose?

```typescript
const renderTo = (canvasId: string) => <A>(r: C.Render<A>, onNone: () => IO.IO<void>): IO.IO<void> =>
  pipe(C.getCanvasElementById(canvasId), IO.chain(O.fold(onNone, flow(C.getContext2D, IO.chain(r)))))
```

gcanti commented on test/Canvas.test.ts:
```diff
@@ -0,0 +1,1330 @@
+      const second = S.point(10, 10)
+
+      // Test
+      pipe(
+        canvas,
+        C.getContext2D,
+        IO.chain(C.fillPath(flow(C.setFillStyle(color), IO.chain(C.moveTo(first)), IO.chain(C.lineTo(second)))))
+      )()
```
this should be

```ts
render(
  C.fillPath(
    pipe(
      C.setFillStyle(color),
      R.chain(() => C.moveTo(first)),
      R.chain(() => C.lineTo(second))
    )
  )
)()
```

I guess

IMax153 replied:
Apologies, missed that one on the initial pass. See https://github.com/gcanti/graphics-ts/pull/7/commits/c71374969b11532dbc227cf7f069517546a0214d

gcanti replied:
NP

gcanti commented:
> I worked on refactoring the tests a bit to make use of the updates to the Canvas and Drawing modules. I figured that users of the library may refer to the tests to gain a better understanding of how graphics-ts functions, so they should probably represent intended use.

:+1:

> One problem that I am having is how I can thread Gradient<A> effects into Render<A>

I would refactor to something like

```ts
// test 'createLinearGradient'

const g: C.Gradient<CanvasGradient> = pipe(
  C.addColorStop(0, acqua),
  R.chain(() => C.addColorStop(0.5, green)),
  R.chain(() => C.addColorStop(1, green))
)

const r: C.Render<CanvasRenderingContext2D> = pipe(
  C.createLinearGradient(x0, y0, x1, y1),
  R.chain((cg) => R.fromIO(g(cg))),
  R.chain(C.setFillStyle),
  R.chain(() => C.fillRect(rect))
)

// Test
render(r)()

// Actual
const testGradient = testCtx.createLinearGradient(x0, y0, x1, y1)
testGradient.addColorStop(0, acqua)
testGradient.addColorStop(0.5, green)
testGradient.addColorStop(1, green)
testCtx.fillStyle = testGradient
testCtx.fillRect(rect.x, rect.y, rect.width, rect.height)
```

> Another question I had is whether we should export a helper function for querying the DOM for a canvas and executing a Render<A> effect

Okay. That sounds good to me. 

Nitpick: I would move the error handler near the source of error, i.e. the canvas id, also maybe we could use a more descriptive name

```ts
const renderTo = (canvasId: string, onCanvasNotFound: () => IO.IO<void>) => <A>(r: C.Render<A>): IO.IO<void> =>
  pipe(C.getCanvasElementById(canvasId), IO.chain(O.fold(onCanvasNotFound, flow(C.getContext2D, IO.chain(r)))))
```

IMax153 committed:
Refactor Gradient effect tests

See https://github.com/gcanti/graphics-ts/pull/7#issuecomment-635070591

Also refactors the last Render effect that was missed in the last pass.

See https://github.com/gcanti/graphics-ts/pull/7#discussion_r431556667

IMax153 committed:
Add renderTo helper function to Drawing module

Additional changes:
- add tests for renderTo
- fix CHANGELOG
- bump TypeScript version to 3.9.3

See https://github.com/gcanti/graphics-ts/pull/7#issuecomment-635070591

gcanti commented:
We can now leverage the new `renderTo` function in both the top examples (in `Canvas.ts` and `Drawing.ts`).

Btw shouldn't `renderTo` go into the `Canvas.ts` module?

IMax153 committed:
Add missing canvas API functions and tests

See CHANGELOG.md for all added functions.

IMax153 commented:
> Btw shouldn't `renderTo` go into the `Canvas.ts` module?

I honestly wasn't sure, but I am happy to move it over to the `Canvas.ts` module. I will also go ahead and add `renderTo` in the examples for the `Drawing.ts` and `Canvas.ts` modules.

I was also reviewing the HTML Canvas API on [MDN](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D) and noticed that we were missing a few `CanvasRenderingContext2D` methods that are not considered experimental. I added those functions, with their associated tests https://github.com/gcanti/graphics-ts/pull/7/commits/213d4ac39705a76ade8f2c1064d63d55b3187aa1. Let me know what you think.

Additionally, I feel that the modules are getting a bit messy. I was going to go ahead and re-organize the modules and test files for clarity (i.e. alphabetize product types, group exports according to functionality), but wanted to ask you before I start moving everything around.

gcanti commented:
go ahead, this is your rewrite and you are doing a wonderful job, I will only intervene if you need some help

IMax153 committed:
Reorganize modules for clarity and to match MDN documentation

IMax153 committed:
Add hex color support to the Color module

IMax153 committed:
Update docs to include hex color documentation

IMax153 commented:
> go ahead, this is your rewrite and you are doing a wonderful job

@gcanti - Thank you very much for those kind words, it means a lot. And for helping me to learn more about functional programming during this rewrite.

I believe this is ready for final review. I made one additional change to the `Color.ts` module and added support for hexadecimal color values since I found myself frequently wanting to use hex values for colors.

IMax153 committed:
Fix links in documentation

IMax153 committed:
Update CHANGELOG with addition of hex color constructor

IMax153 committed:
Export shape from index file

IMax153 commented:
@gcanti - As I mentioned before, I have been using the library for a personal project. My project involves binding event handlers to the canvas to allow for event-based animations. 

I am curious what your thoughts are on adding the following to the `Canvas.ts` module to make it easier for users to do the same.

```typescript
export type Handler<E extends Event> = (e: E) => void

export const bind: <K extends keyof HTMLElementEventMap>(
  type: K,
  f: Handler<HTMLElementEventMap[K]>
) => Html<HTMLCanvasElement> = (t, f) => (c) => () => {
  c.addEventListener(t, f)
  return c
}
```

gcanti commented:
@IMax153 seems sensible

IMax153 committed:
Add bind function to Canvas module

The bind function allows a consumer of the library to easily bind
an event handler to the managed HTMLCanvasElement.

See https://github.com/gcanti/graphics-ts/pull/7#issuecomment-636562547

IMax153 committed:
Add tests for bind function

IMax153 committed:
Remove redundant setup code from Canvas test file

IMax153 committed:
Add contructors for working with angles

Users can now freely specify angles in either degrees or radians

IMax153 commented:
@gcanti  - The last major pain point I came across when using the library was wanting to freely switch between specifying angles in degrees and radians. [This](https://github.com/gcanti/graphics-ts/pull/7/commits/f273099e1a669b0cd8d8efbdf36c9265785e519e) last commit adds constructors which allows for users to specify angles as `degrees` or `radians`, and automates resolving the angle to a radian value for use with the canvas context.

I think the rewrite is in a good place at this point and I don't have much more to add after having experimented with the library. I think this PR is done 👍.

IMax153 committed:
Add anticlockwise to arc constructor on Canvas

gcanti commented:
> I think this PR is done

Great! Feel free to merge when you are done

IMax153 commented:
@gcanti - Thanks for the opportunity to work on this update. Was a lot of fun and I learned a lot! Looking forward to using the lib in practice.

I was thinking about writing a blog post on dev.to about my experience learning about the library and showcasing some of its features. Just wanted to check that it was alright with you before I went ahead with that.

gcanti commented:
> Thanks for the opportunity to work on this update. Was a lot of fun and I learned a lot! Looking forward to using the lib in practice.

Thanks to you, you did an awesome job.

> I was thinking about writing a blog post on dev.to about my experience learning about the library and showcasing some of its features. Just wanted to check that it was alright with you before I went ahead with that.

Sure! Looking forward to read it

gcanti commented:
Released

IMax153 commented:
> Released

Thanks for letting me know @gcanti! This is excellent! 👍 

I just wanted to let you know about two issues:

1. In the [`import-path-rewrite](https://github.com/gcanti/import-path-rewrite) package, the `packages` array does not contain `graphics-ts` so I don't believe the paths will be correctly rewritten
2. I believe the docs for `graphics-ts` still need to be published since the link in the `README` currently returns a 404

Also, in case you would like to review/make edits - the blog post is published [here](https://dev.to/imax153/graphics-ts-functional-bindings-for-the-html-5-canvas-api-1832)

gcanti commented:
> In the [import-path-rewrite](https://github.com/gcanti/import-path-rewrite) package, the packages array does not contain `graphics-ts` so I don't believe the paths will be correctly rewritten

We only need to rewrite the `fp-ts-contrib` / `fp-ts` imports.

> I believe the docs for graphics-ts still need to be published since the link in the README currently returns a 404

Ops... you are right, they should be up now.
